xwiki-platform-core/xwiki-platform-localization/xwiki-platform-localization-rest/src/main/java/org/xwiki/localization/internal/DefaultLocalizationSource.java

```diff
@@ -0,0 +1,113 @@
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.localization.internal;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.LocaleUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.localization.LocalizationContext;
+import org.xwiki.localization.LocalizationManager;
+import org.xwiki.localization.Translation;
+import org.xwiki.localization.rest.LocalizationSource;
+import org.xwiki.rest.XWikiRestComponent;
+
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Default implementation of {@link LocalizationSource}.
+ *
+ * @version $Id$
+ * @since 13.3RC1
+ */
+@Component
+@Named("org.xwiki.localization.internal.DefaultLocalizationSource")
+@Singleton
+public class DefaultLocalizationSource implements LocalizationSource, XWikiRestComponent
+{
+    @Inject
+    private LocalizationContext localizationContext;
+
+    @Inject
+    private LocalizationManager localizationManager;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public Response translations(String wikiId, String localeString, String prefix, List<String> keys)
+    {
+        XWikiContext xWikiContext = this.contextProvider.get();
+
+        // Save the current context wiki id. It will be restored on the method is finished.
+        String oldWikiId = xWikiContext.getWikiId();
+        try {
+            // Set the requested wiki id.
+            xWikiContext.setWikiId(wikiId);
+
+            // Resolve the local.
+            Locale locale;
+            if (localeString != null) {
+                locale = LocaleUtils.toLocale(localeString);
+            } else {
+                locale = this.localizationContext.getCurrentLocale();
+            }
+
+            // Computes the prefix.
+            String cleanPrefix = Objects.toString(prefix, "");
```

Comment:
You're doing that because prefix is nullable? It's not indicated in javadoc AFAICS.

Reply:
Exactly, and all jaxrs query params are nullable so I didn't repeat it in the javadoc, but explicit is always better, I'll update the javadoc.

Reply:
> all jaxrs query params are nullable

so you're missing a check for keys below no?

Reply:
(and wikiId?)

Reply:
wikiId is mandatory since it's a path param. I have a doubt regarding `keys`. when no `key` parameter is provided, it is initialized with and list containing a single null element, but indeed in this case the response is `{"null": null}` which is not what we want, fixing.

Reply:
Missing key parameter fix https://github.com/xwiki/xwiki-platform/pull/1595/files#diff-118a400f3886da870ebde91eda0d6269f44a19975059f7d189964ac02c48d623R94
Improved javadoc https://github.com/xwiki/xwiki-platform/pull/1595/files#diff-d3378dd806e951aaca607e611516c5e5bc45d33347bba4e1eaf0a8909578c803R42-R57